### PR TITLE
Removing the master-slave terminology

### LINKED ISSUE #35: Remove the master-slave terminology
Remove the master-slave terminology

Criteria

- Rename master for main
- Rename slave for subordinate

This will be a breaking change, since it will change the interface of the software. (the name of the utilities, among others). A major release should follow.

### PR DESCRIPTION
**Pull request by bopopescu**
For diversity reasons, it would be nice to try to avoid 'master' and 'slave' terminology in this repository which can be associated to slavery. The master-slave terminology could be problematic for people in several countries which has the history of slavery like Romania, USA and many others. Thank you for considering the proposal. Let me know if any changes in the PR are needed, I would be happy to implement them.

### CODE REVIEW

**Comment by aalex** [comment]
Hi! Thank you for your proposal. I was thinking about the same thing, so I would be happy to squash and rebase your PR. Let me take some time to inspect it. 

I was going to port Lunch to Python 3 within the next few days, so your contribution might very well end up in the next upcoming official release of Lunch. I am a bit surprise to see that this great software is being used by other people. It used to be packaged for Debian, but it was no longer building, probably due to the drop of Python 2.

So, I should check your PR pretty quickly.

**Comment by aalex** [comment]
@bopopescu It seems to me that this PR should be against the develop branch. It also contains some commits that should not be in. Perhaps I can help rebasing this branch onto dev, so that it's only one commit against the right branch, if you wish. Can you change the branch to which it's against, perhaps? And then I can help with the rebase, if needeed.

**Review by aalex — CHANGES_REQUESTED** [review]
- This PR should be against the develop branch.
- It should not contain old commits. (let's rebase it onto develop)

**Comment by djiamnot** [comment]
This PR seems to address two different topics: one refers to git branch names, another to terminology used to describe some internal logic. I am not sure if `main` and `subordinate` actually work as replacement terms in the latter case.

**Comment by aalex** [comment]
> This PR seems to address two different topics: one refers to git branch names, another to terminology used to describe some internal logic. I am not sure if `main` and `subordinate` actually work as replacement terms in the latter case.

Hello @djiamnot !

I think that this PR is only about renaming lunch-master and lunch-slave to lunch-main and lunch-subordinate. I don't think it has to do anything with Git branch naming. (unless I am mistaken)

But yes: I have been thinking about, and I agree with you that the name main and subordinate might not be the best choice. I think "main" confuses me with "\_\_main\_\_", among other things, and "subordinate" sounds way too complicated for me. I like when things are named in a simple and explicit way. Let's think better names, if possible. Actually, discussing which names to rename those to should be done prior to doing pull requests.

As I wrote earlier, this is a pretty good time to do that change, if possible, because I am about to port Lunch to Python 3, since I need to use it for a project - but this is for quite soon. 

About a good name, I have used a lot Docker recently, and I am using Kubernetes as well. I wonder if we can get inspired from the terms used in the other devops deployment software. (also: Puppet, Vagrant, Ansible, Chef, Terraform, etc.) 

Here are few quick options to replace the term "master":

- controller
- manager

And for the term "slave":

- lunchlet
- service
- pod
- agent
- node
- remote
- proxy

**Comment by djiamnot** [comment]
oh, yes, sorry, I was thrown off by changes in `CONTRIBUTING` and `NEWS` and the fact that this PR is based on the `develop` branch, asking to be merged into `master`. 

I'm not sure if I prefer _controller_ or _manager_, I let you, @aalex pick the term that resonates with you better. As for "slave", I like "lunchlet", actually. "pod", "agent" and "node" seem quite fitting, too.

**Comment by aalex** [comment]
Manager + Agent

Le dim. 26 juill. 2020 19 h 40, Michal Seta <notifications@github.com> a
écrit :

**Comment by karenetheridge** [comment]
> I am a bit surprise to see that this great software is being used by other people.

It looks like this user has sent thousands of other similar pull requests in the past month, so I suspect they just did a search across all of github, and these changes were generated by a script, without much human review.

**Comment by aalex** [comment]
> > I am a bit surprise to see that this great software is being used by other people.
> 
> It looks like this user has sent thousands of other similar pull requests in the past month, so I suspect they just did a search across all of github, and these changes were generated by a script, without much human review.

Ah! That explains a lot.

I will make that low-priority, then, because I need to release a new version soon - I will concentrate on the bug fixes I need in the short term. 

If someone wants to create another PR using the nomenclature that we've discussed, I could accept it.

The names chosen are: manager and agent.

File: lunch/states.py [hunk 1]
```diff
@@ -19,13 +19,13 @@
...
 """
-Constants for the states of the slave processes
+Constants for the states of the subordinate processes
 """
 STATE_STARTING = "STARTING"
 STATE_RUNNING = "RUNNING"
 STATE_STOPPING = "STOPPING"
 STATE_STOPPED = "STOPPED"
-STATE_NOSLAVE = "NOSLAVE" # for master only
+STATE_NOSLAVE = "NOSLAVE" # for main only
```
Comment (aalex) [inline lunch/states.py:28]: The value and name of this constant should be changed as well.

File: lunch/__init__.py [hunk 2]
```diff
@@ -1,7 +1,6 @@
 """Lunch: Distributed process launcher"""
-__version__ = "0.6.0"
-DEFAULT_LOG_DIR = "/tmp/"
-DEFAULT_PID_DIR = "/tmp/"
+__version__ = "0.7.0"
+DEFAULT_LOG_DIR = "/tmp"
+DEFAULT_PID_DIR = "/tmp"
```
Comment (aalex) [inline lunch/__init__.py:4]: It's probably a good idea to remove those slashes.

File: examples/using_the_library/simple.py [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
     # XXX add_command here
-    m = master.Master(log_dir=log_dir, pid_file=pid_file)
+    m = main.Main(log_dir=log_dir, pid_file=pid_file)
     m.add_command(commands.Command("xeyes", identifier="xeyes"))
     m.add_command(commands.Command("xlogo", identifier="xlogo"))
     m.add_command(commands.Command("xcalc", identifier="xcalc"))
     def _test():
         print("Adding one more!")
-        m.add_command(commands.Command("xeyes"))
+        m.add_command(commands.Command("xeyes -geometry %dx%d" % (random.randint(20, 500), random.randint(20, 500))))
```
Comment (aalex) [inline examples/using_the_library/simple.py:26]: Nice contribution!

**Comment by aalex** [comment]
I think that I could rebase this commit onto the `develop` branch, and change it to keep only two minor edits:

- removing the slashes at the end of some paths
- a random position for the xeyes in an example

And then, we will do another PR to rename the master and slave to something else.

**Pull request closed without merging**